### PR TITLE
Implement non application error codes fix #895

### LINKED ISSUE #895: Non-application-protocol error codes
When there's a non-application-protocol error for a HTTP request or WebSocket connection today k6 sets a status of 0 (property `status` in response object and metric output tags) and adds a Go stdlib message to the `error` property in the response object as well as the metrics output tags.

This presents a number of problems when trying to analyze the data, whether manually or computer-aided:

- You need to parse text to find the actual type of error (DNS? TCP? connect fail? recv/send fail? etc.)
    - If you're not familiar with the Go stdlib error messages (most won't be) it can be difficult to understand what the error actually is
- There's no straightforward way to aggregate the errors as some of the `error` messages contain randomly selected src and dest TCP port numbers, meaning each message is unique although reporting the same issue

I therefore propose we introduce a new property `error_code`. A pre-defined numeric value that we map to the list of possible errors that k6 can run into. This solves the problems mentioned above; it will make it much easier to aggregate and present error stats and we can provide much more end-user digestible error messages.

I think we should also define ranges of values for different categories/layers of errors like: DNS, TCP, TLS, HTTP etc.

The gist of the proposal:
- Introduce a `error_code` property to the Response object [1].
- Introduce a `error_code` tag in the metrics output
- Use the following ranges of codes for specific categories/layers:
    - 0: no error, everything's OK!
    - 1...99: Reserved
    - 100-199: DNS (needs to be broken down into specific DNS error codes that k6 can report)
    - 200-299: TCP (needs to be broken down into specific TCP error codes that k6 can report)
    - 300-399: TLS (needs to be broken down into specific TLS error codes that k6 can report)
    - 400-499: HTTP client side errors (simple maps to HTTP 4xx status codes)
    - 500-599: HTTP server side errors (simple maps to HTTP 5xx status codes)
    - 600-699: WebSocket
- Set the `error` Response property and metrics tag to a new, user friendlier error message written by us for each of the possible error codes

[1] - https://docs.k6.io/docs/response-k6http
[2] - https://github.com/loadimpact/k6/issues/66 (the original issue regarding this topic, with the discussion that resulted in the current solution with `status: 0` and `error: message`)

### PR DESCRIPTION
**Pull request by MStoykov**

### CODE REVIEW

File: js/modules/k6/http/error_codes.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+		if e.Net != "tcp" && e.Net != "tcp6" {
+			// TODO: figure out how this happens
+			return defaultNetNonTCPErrorCode
+		}
+		if e.Op == "write" {
+			switch e.Err.Error() {
+			case syscall.ECONNRESET.Error():
+				return tcpResetByPeerErrorCode
+			case syscall.EPIPE.Error():
+				return tcpBrokenPipeErrorCode
```
[the remark this answers is not in the document]
Comment (MStoykov) [reply js/modules/k6/http/error_codes.go]: I don't really think I want to have a default defaultTCPWriteErrorCode , I don't really think it adds more than defaultTCPErrorCode alone

File: js/modules/k6/http/error_codes.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		case "no such host": // defined as private in the go stdlib
+			return dnsNoSuchHostErrorCode, dnsNoSuchHostErrorCodeMsg
+		default:
+			return defaultDNSErrorCode, ""
+		}
+	case netext.BlackListedIPError:
+		return blackListedIPErrorCode, blackListedIPErrorCodeMsg
+	case *http2.GoAwayError:
+		// TODO: Add different error for all errcode for goaway
+		return unknownHTTP2GoAwayErrorCode + http2ErrCodeOffset(e.ErrCode), fmt.Sprintf(http2GoAwayErrorCodeMsg, e.ErrCode)
```
Comment (na--) [inline js/modules/k6/http/error_codes.go]: Not sure if `unknown` is the best word choice here and below, since we likely have the specific error codes and know the actual errors. `base` or something like that seems better, i.e. `baseHTTP2GoAwayErrorCode`, `baseHTTP2StreamErrorCode`, etc.

File: js/modules/k6/http/error_codes.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		switch e.Err {
+		case "no such host": // defined as private in the go stdlib
+			return dnsNoSuchHostErrorCode, dnsNoSuchHostErrorCodeMsg
+		default:
+			return defaultDNSErrorCode, ""
+		}
+	case netext.BlackListedIPError:
+		return blackListedIPErrorCode, blackListedIPErrorCodeMsg
+	case *http2.GoAwayError:
+		// TODO: Add different error for all errcode for goaway
```
Comment (na--) [inline js/modules/k6/http/error_codes.go]: This `TODO` and the similar ones below seem done?

File: js/modules/k6/http/error_codes.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+func http2ErrCodeOffset(code http2.ErrCode) errCode {
+	if code > http2.ErrCodeHTTP11Required {
+		return 0
+	}
+	return 1 + errCode(code)
+}
+
+// returns the errorCode and a specific error message for given error. If errror message is empty
+// than the original error string should be used
+func errorCodeForError(err error) (errCode, string) {
```
[the remark this answers is not in the document]
Comment (MStoykov) [reply js/modules/k6/http/error_codes.go]: I agree with all of those except that we are not there yet and at this point I think it is better if we don't optimise for what is needed in the future. It is specifically this that I moved it to a separate file and is totally separate from everything, but as is currently not required by anything I will be making it a separate module with the idea that one day someone will remember to use it ...

File: js/modules/k6/http/error_codes.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		// TODO: Add different error for all errcode for goaway
+		return unknownHTTP2GoAwayErrorCode + http2ErrCodeOffset(e.ErrCode), fmt.Sprintf(http2GoAwayErrorCodeMsg, e.ErrCode)
+	case *http2.StreamError:
+		// TODO: Add different error for all errcode for stream error
+		return unknownHTTP2StreamErrorCode + http2ErrCodeOffset(e.Code), fmt.Sprintf(http2StreamErrorCodeMsg, e.Code)
+	case *http2.ConnectionError:
+		return unknownHTTP2ConnectionErrorCode + http2ErrCodeOffset(http2.ErrCode(*e)), fmt.Sprintf(http2ConnectionErrorCodeMsg, http2.ErrCode(*e))
+	case *net.OpError:
+		if e.Net != "tcp" && e.Net != "tcp6" {
+			// TODO: figure out how this happens
```
[the remark this answers is not in the document]
Comment (MStoykov) [reply js/modules/k6/http/error_codes.go]: Have no idea - left it as a note for when/if we see such an error - as far as I know it is not possible for us currently to get such an error which is why it has separate error_code so we can see it ;)

File: js/modules/k6/http/response.go [hunk 6]
```diff
@@ -66,12 +66,32 @@ type Response struct {
...
+		errorMsg = err.Error()
+	}
+	res.Error = errorMsg
+}
+
+// This should be used instead of setting Error as it will correctly set ErrorCode as well
+func (res *Response) setStatusCode(statusCode int) {
+	res.Status = statusCode
+	if statusCode >= 400 && statusCode < 600 {
+		res.ErrorCode = 1000 + statusCode // TODO: Maybe this should not add 1000?
```
[the remark this answers is not in the document]
Comment (MStoykov) [reply js/modules/k6/http/response.go]: :+1:

File: js/modules/k6/http/error_codes.go (hunk 2 above)
Comment (MStoykov) [reply js/modules/k6/http/error_codes.go]: Yeah but this specific one will be used in case of unknown [http2 ErrCode](https://godoc.org/golang.org/x/net/http2#ErrCode) which seems unlikely but I decided to be confidence instead of getting error 1721 and than having to figure out it is actually a GoAway with some strange ErrCode. I also don't like the name though ... maybe default is fine after alll and just the other ones are specifics ?

File: js/modules/k6/http/error_codes.go (hunk 3 above)
Comment (MStoykov) [reply js/modules/k6/http/error_codes.go]: :+1:  yeah apparently I forgot them :(

File: js/modules/k6/http/error_codes.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (na--) [reply js/modules/k6/http/error_codes.go]: Yeah, but we know that we'd like to soon reuse this code for other existing functionality (websocket errors, when we fix their metrics emission - last part of https://github.com/loadimpact/k6/issues/885#issuecomment-450642259), and things like gRPC and other protocols, which we'd like to implement as soon as possible. So it's not refactoring for some nebulous future, it'd be needed for stuff that we plan to soon implement. Even worse, other users may want to implement them, and having this here will be in the way...

Since if we or any users want to implement one of those things, and this code is still in the `js/http` package, it would likely first have to moved in a new package in one PR, with the actual new feature in another PR, which seems like a lot of wasted effort and an obstacle to contributors. Especially if we do any changes here in the meantime, that depend on the local package nature, and which would have to be fixed before refactoring the package in the future.

[the remark this answers is not in the document]
Comment (na--) [reply js/modules/k6/http/error_codes.go]: Also, I think the philosophical point is somewhat moot, since it seems like we may actually need to do this now anyway... I decided to look into how we can add the `error_code` metric tag, since I didn't see how we add the `error` tag to HTTP metrics currently. I didn't find it in this package, it happens here instead: https://github.com/loadimpact/k6/blob/master/lib/netext/transport.go#L57-L62

I think the `error` tag got moved there in the fix to properly measure HTTP requests when there are redirects. And it makes some sense, since an HTTP request can have a successful redirect (no error) to a URL that returns an error. And we need to emit both of those sets of metrics. But that also means that the `error` metric tag won't be fixed by this PR, it will still contain highly variable strings...

[the remark this answers is not in the document]
Comment (na--) [reply js/modules/k6/http/error_codes.go]: Regarding that last observation: there should also probably be a test in this PR that actually checks this at the engine level, by making one or more broken HTTP requests, capturing the metrics and seeing if the `error` and `error_code` tags look like we want them to look

[the remark this answers is not in the document]
Comment (MStoykov) [reply js/modules/k6/http/error_codes.go]: Testing it that way was my initial idea but decided that I will do it only for one or two errors that will be easier to reproduce :) like blacklist Ip and let say an unresolvable address as reproducing the majority of the errors will be kind of hard :(

File: lib/netext/httpext/transport.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/netext"
+	"github.com/loadimpact/k6/stats"
+	"github.com/pkg/errors"
+)
+
+// transport is an implemenation of http.RoundTripper that will measure different metrics for each
+// roundtrip
+type transport struct {
```
Comment (na--) [inline lib/netext/httpext/transport.go]: Still not sure why you un-exported this, especially after you bothered to properly comment all of the methods :smile:

File: lib/netext/httpext/request.go [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+
+	if resErr != nil {
+		// Do *not* log errors about the contex being cancelled.
+		select {
+		case <-ctx.Done():
+		default:
+			state.Logger.WithField("error", resErr).Warn("Request Failed")
+		}
+
+		if preq.Throw || resp.Error == "" {
```
Comment (na--) [inline lib/netext/httpext/request.go]: I am confused, why did you add `|| resp.Error == ""` to this check?

File: lib/netext/httpext/request.go [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+		resp.ErrorCode = int(tracerTransport.errorCode)
+		if err != nil {
+			// Do *not* log errors about the contex being cancelled.
+			select {
+			case <-ctx.Done():
+			default:
+				state.Logger.WithField("error", res).Warn("Digest request failed")
+			}
+
+			if preq.Throw || resp.Error == "" {
```
Comment (na--) [inline lib/netext/httpext/request.go]: Same confusion about adding `|| resp.Error == ""` here.

File: lib/netext/httpext/transport.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	ctx := req.Context()
+	tracer := Tracer{}
+	reqWithTracer := req.WithContext(httptrace.WithClientTrace(ctx, tracer.Trace()))
+
+	resp, err := t.roundTripper.RoundTrip(reqWithTracer)
+	trail := tracer.Done()
+	if err != nil {
+		t.errorCode, t.errorMsg = errorCodeForError(err)
```
Comment (na--) [inline lib/netext/httpext/transport.go]: I think that by setting the values of `t.errorCode` and `t.errorMsg` only in the `if err != nil` branch (and for the 400+ status codes) and not resetting them in the beginning of `RoundTrip()`, you could potentially have wrong values when we (in the `ntlm`/`digest` auths) or the stdlib reuse the same `RoundTripper` for multiple round trips.

File: js/modules/k6/http/request_test.go [hunk 11]
```diff
@@ -450,6 +450,122 @@ func TestRequestAndBatch(t *testing.T) {
 		assert.Error(t, err)
 	})
 
+	t.Run("ErrorCodes", func(t *testing.T) {
+		defer func(value null.Bool) {
+			state.Options.Throw = value
```
[the remark this answers is not in the document]
Comment (MStoykov) [reply js/modules/k6/http/request_test.go]: I was wondering the same thing at the begging and then never did it - will do it now.

File: js/modules/k6/http/response.go [hunk 12]
```diff
@@ -22,102 +22,41 @@ package http
...
-	TLSHandshaking float64 `json:"tls_handshaking"`
-	Sending        float64 `json:"sending"`
-	Waiting        float64 `json:"waiting"`
-	Receiving      float64 `json:"receiving"`
-}
-
 // Response is a representation of an HTTP response to be returned to the goja VM
-type Response struct {
-	ctx context.Context
+// TODO: refactor after https://github.com/dop251/goja/issues/84
```
[the remark this answers is not in the document]
Comment (MStoykov) [reply js/modules/k6/http/response.go]: Definitely not in this PR - as it will mean upgrading goja in this one as well and this one already has a bunch of non related things in it. I added https://github.com/loadimpact/k6/issues/946

File: lib/netext/httpext/request.go (hunk 8 above)
Comment (MStoykov) [reply lib/netext/httpext/request.go]: As per the offline discussion we had: if we get an error that is not from the RoundTripper we should throw an exception in goja this is basically the same behaviour as if `throw` is set to true. I check `resp.Error` as I should have previously set it to the error from the RoundTripper if it is empty but we have an error it means the error is not from the RoundTripper and we want to throw an exception in goja.

File: lib/netext/httpext/transport.go (hunk 7 above)
Comment (MStoykov) [reply lib/netext/httpext/transport.go]: for me being exported is only acceptable when it will be used outside of the package .. it isn't any more and ... I doubt it will ever be again ;) if it becomes we can always export it again

File: js/modules/k6/http/response.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (na--) [reply js/modules/k6/http/response.go]: :+1:

File: lib/netext/httpext/transport.go (hunk 10 above)
Comment (MStoykov) [reply lib/netext/httpext/transport.go]: We never do reuse the RoundTripper `https://github.com/loadimpact/k6/pull/907/files#diff-aae3e0dbcd566069bafcba0f2c08daafR155` I suppose I can nil them at the begining but I don't know if it won't break somewhere else in either case

Comment (na--) [reply lib/netext/httpext/transport.go]: Check just a few lines below the code you linked :smile: 
```go
transport = ntlmssp.Negotiator{
	RoundTripper: tracerTransport,
}
```
NTLM negotiation and digest authentication reuse the same transport, they even have some expected 4xx requests...

File: lib/netext/httpext/request.go (hunk 8 above)
Comment (na--) [reply lib/netext/httpext/request.go]: Can you add something like this (maybe with a few commas :wink: ) as a comment inside of the code, since it totally isn't obvious from just reading the code.

File: js/modules/k6/http/request_test.go [hunk 13]
```diff
@@ -469,7 +585,16 @@ func TestRequestAndBatch(t *testing.T) {
 				state.CookieJar = cookieJar
 				_, err = common.RunString(rt, sr(`
 				let res = http.request("GET", "HTTPBIN_URL/cookies/set?key=value", null, { redirects: 0 });
-				if (res.cookies.key[0].value != "value") { throw new Error("wrong cookie value: " + res.cookies.key[0].value); }
```
[the remark this answers is not in the document]
Comment (MStoykov) [reply js/modules/k6/http/request_test.go]: No Idea - the added codes are from long time ago I just returned them and maybe ... I deleted that one by accident - returning it now

**Comment by MStoykov** [comment]
About removing `error` from the default tags: removing it will prevent us from more easily identifying currently non captured errors and to give them codes. So I am all for leaving it
